elyra/contents/file_parser.py

```diff
@@ -0,0 +1,197 @@
+#
+# Copyright 2018-2021 Elyra Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+import nbformat
+import re
+
+from abc import ABC, abstractmethod
+
+from typing import Any, Type, TypeVar, List, Dict
+from traitlets import log
+from ..util import get_expanded_path, get_absolute_path
+
+# Setup forward reference for type hint on return from class factory method.  See
+# https://stackoverflow.com/questions/39205527/can-you-annotate-return-type-when-value-is-instance-of-cls/39205612#39205612
+F = TypeVar('F', bound='FileParser')
+
+
+class FileParser(ABC):
+
+    @classmethod
+    def get_instance(cls: Type[F], **kwargs: Any) -> F:
+        """Creates an appropriate subclass instance based on the extension of the filepath"""
+        filepath = kwargs['filepath']
+
+        if '.ipynb' in filepath:
+            return NotebookFileParser(filepath)
+        elif '.py' in filepath:
+            return PythonFileParser(filepath)
+        elif '.r' in filepath:
+            return RFileParser(filepath)
+        else:
+            raise ValueError('Unsupported file type: {}'.format(filepath))
+
+    _type = None
+
+    def __init__(self, operation_filepath):
+        self._operation_filepath = operation_filepath
+        self._parser = None
+
+        # Consider adding log messages
+        self.log = log.get_logger()
+
+    @property
+    @abstractmethod
+    def type(self):
+        pass
+
+    @property
+    def parser(self):
+        if not self._parser:
+            raise ValueError(f'Could not find appropriate language parser for {self.operation_filepath}')
+        return self._parser
+
+    @property
+    def operation_filepath(self):
+        # Retrieve the absolute path
+        root_dir = get_expanded_path()
+        abs_path = get_absolute_path(root_dir, self._operation_filepath)
+
+        if not os.path.exists(abs_path):
+            raise FileNotFoundError(f'Could not find {abs_path}')
+        if not os.path.isfile(abs_path):
+            raise ValueError(f'Not a file: {abs_path}')
+
+        self._operation_filepath = abs_path
+        return self._operation_filepath
+
+    @abstractmethod
+    def get_next_code_chunk(self) -> List[str]:
```

Comment:
Since we now have "base" implementation in this method, it should not be marked abstract.  If the abstract property `type` also ends up going away, we can look into other subclasses, [`LoggingConfigurable`](https://github.com/ipython/traitlets/blob/2bb2597224ca5ae485761781b11c06141770f110/traitlets/config/configurable.py#L432) might be one, then we get the inherited logger.

Reply:
Looks like type is no longer used anywhere. Actually, I'm not sure we used it before either ha. Anyway, I'll go ahead and use `LoggingConfigurable` then since we no longer have any abstract methods/properties